fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java

```diff
@@ -1399,66 +1448,73 @@ private static void parseInclusionParameter(Class<?> resourceType, FHIRSearchCon
-            searchParameterName       = inclusionValueParts[1];
+            joinResourceType = inclusionValueParts[0];
+            searchParameterName = inclusionValueParts[1];
             searchParameterTargetType = inclusionValueParts.length == 3 ? inclusionValueParts[2] : null;
 
-            // Ensure that the Inclusion Parameter being parsed is a valid search parameter of type 'reference'.
-            searchParm                = getSearchParameter(joinResourceType, searchParameterName);
-            if (searchParm == null) {
-                String msg = "Undefined Inclusion Parameter: " + inclusionValue;
-                if (lenient) {
-                    // TODO add this to the list of supplemental warnings?
-                    log.fine(msg);
-                    continue;
-                } else {
-                    throw SearchExceptionUtil.buildNewInvalidSearchException(msg);
-                }
+            // For _include parameter, join resource type must match resource type being searched
+            if (SearchConstants.INCLUDE.equals(inclusionKeyword) && !joinResourceType.equals(resourceType.getSimpleName())) {
+                throw SearchExceptionUtil.buildNewInvalidSearchException(
+                        "The join resource type must match the resource type being searched.");
             }
-            if (!searchParm.getType().getValue().equals("reference")) {
-                throw SearchExceptionUtil
-                        .buildNewInvalidSearchException("Inclusion Parameter must be of type 'reference'. "
-                                + "The passed Inclusion Parameter is of type: " + searchParm.getType().getValue());
+
+            // For _revinclude parameter, target resource type, if specified, must match resource type being searched
+            if (SearchConstants.REVINCLUDE.equals(inclusionKeyword) && searchParameterTargetType != null
+                    && !searchParameterTargetType.equals(resourceType.getSimpleName())) {
+                throw SearchExceptionUtil.buildNewInvalidSearchException("The search parameter target type must match the resource type being searched.");
             }
 
-            if (inclusionKeyword.equals(SearchConstants.INCLUDE)) {
-                newInclusionParms =
-                        buildIncludeParameter(resourceType, joinResourceType, searchParm, searchParameterName,
-                                searchParameterTargetType);
-                context.getIncludeParameters().addAll(newInclusionParms);
+            // Ensure that the Inclusion Parameter being parsed is a valid search parameter of type 'reference'.
+            Map<String, SearchParameter> searchParametersMap;
+            if (SearchConstants.WILDCARD.equals(searchParameterName)) {
+                searchParametersMap = getInclusionWildcardSearchParameters(resourceType.getSimpleName(), joinResourceType, searchParameterTargetType, inclusionKeyword);
```

Comment:
What happens when it's set to lenient? e.g. do we fail it? look at the behavior on line 1477

Reply:
In the case of a wildcard search, if no valid matching search parameters are found, then we simply don't generate an _include or _revinclude clause - no exception is thrown. Should an exception be thrown in that case (if non-lenient)? I can certainly add a log entry indicating no valid search parms were found.

Reply:
Maybe we just log out, if it is set to not lenient?

Reply:
Yeah, will just log it.